pkg/controller/atlascluster/advanced_cluster.go

```diff
@@ -0,0 +1,124 @@
+package atlascluster
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"go.mongodb.org/atlas/mongodbatlas"
+	"go.uber.org/zap"
+
+	mdbv1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/workflow"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/util/compat"
+)
+
+func (r *AtlasClusterReconciler) ensureAdvancedClusterState(ctx *workflow.Context, project *mdbv1.AtlasProject, cluster *mdbv1.AtlasCluster) (*mongodbatlas.AdvancedCluster, workflow.Result) {
```

Comment:
this function is very similar to the regular cluster. There are changes to all of the types to the various functions (api client, merge and clean functions and return type)

I looked into defining an interface, but it ended up being messier than I had hoped.

Reply:
We can try to refactor this again after Go 18, some of the added features might be useful here.

Reply:
yes this was actually a perfect candidate for generics, I was thinking about this when I was writing it!